Merge SSV -> ETH fees & update EB

mtabasco opened the pull request.

venimir-ssv committed:
clusters::registration:eth storage added

venimir-ssv committed:
clusters::registration:refactored

venimir-ssv committed:
clusters::remove:refactored

venimir-ssv committed:
clusters::liquidate:refactored, liquidateSSV added

venimir-ssv committed:
clusters::reactivate:refactored for eth migration

venimir-ssv committed:
clusters::deposit:refactored for eth migration

venimir-ssv committed:
clusters::withdraw:refactored for eth migration

venimir-ssv committed:
operators::library:refactored for eth migration

venimir-ssv committed:
operators::registerOperator:refactored for eth migration

venimir-ssv committed:
operators::removeOperator:refactored for eth migration remove operator ssv function added for backward, clusters missing functions added

venimir-ssv committed:
operators::declareOperatorFee:refactored for eth migration

venimir-ssv committed:
operators::reduceOperatorFee:refactored for eth migration

venimir-ssv committed:
operators::withdraw:refactored for eth migration

venimir-ssv committed:
SSVDAO:refactored for eth migration

venimir-ssv committed:
reentrancy guard added for eth payments

venimir-ssv committed:
migrate to eth operator added

venimir-ssv committed:
migrateClusterToETH added wip

venimir-ssv committed:
ensureETHDefaults added

venimir-ssv committed:
obsolate code removed

venimir-ssv committed:
compilation errors fixed

venimir-ssv committed:
updateClusterOperatorsSSV added

venimir-ssv committed:
ClusterMigratedToETH event added

venimir-ssv committed:
removeValidator nonReentrant modifier removed

venimir-ssv committed:
ssv dao update during migration added

venimir-ssv committed:
withdrawAllVersionOperatorEarnings added

venimir-ssv committed:
ensureETHDefaults refactored

venimir-ssv committed:
Add legacy SSV views, dual withdraw helpers, and bump version to v1.3.0

venimir-ssv committed:
Wire SSVNetworkViews to new SSV/ETH view helpers

venimir-ssv committed:
migrateOperator to ETH refactored

venimir-ssv committed:
reentracy changed to upgradable

venimir-ssv committed:
Initialize reentrancy guard in proxy for delegatecall modules

venimir-ssv committed:
Unify reentrancy guard at proxy and fix ETH/SSV accounting mismatches

yurii-ssv committed:
feat: persist ssv/eth balance checks

yurii-ssv committed:
fix: ssv/eth natspec inconsistency

yurii-ssv committed:
fix: add ethSnapshot check in `checkOwner`

yurii-ssv committed:
chore: fix typos

venimir-ssv committed:
settle SSV snapshot before migrate ETHDefaults, msg.value fixed

venimir-ssv committed:
update SSV before ensureETHDefaults

venimir-ssv committed:
update snapshor on registration removed

venimir-ssv committed:
increase check added

mtabasco committed:
feat:phase 1 - storage

mtabasco committed:
phase 3 - clusters + dao (wip)

liorrutenberg commented on contracts/modules/SSVDAO.sol:
```diff
@@ -75,4 +103,41 @@ contract SSVDAO is ISSVDAO {
+        seb.ebRoots[blockNum] = merkleRoot;
+        seb.latestCommittedBlock = blockNum;
+
+        emit RootCommitted(merkleRoot, blockNum, block.timestamp);
```
@mtabasco @yurii-ssv We need to check if we have quorum of 3/4 or based on delegated vote
and only if it reached quorum to fire RootCommitted event.

yurii-ssv committed:
feat: dao vunits calculation helpers

yurii-ssv committed:
feat: add cluster struct to clusterUpdated event

yurii-ssv committed:
chore: markup & helpers for daoVUnits calculation

yurii-ssv committed:
feat: eb snapshot updates for eth & ssv

yurii-ssv committed:
chore: change init call to 2step upgradeable

yurii-ssv committed:
feat: draft root voting

yurii-ssv committed:
fix: replace root with key

venimir-ssv committed:
cleanup comment

venimir-ssv committed:
fix: align ClusterBalanceUpdated event signature with other cluster events

venimir-ssv committed:
Merge pull request #324 from ssvlabs/fix/cluster-balance-updated-event-order

fix: align ClusterBalanceUpdated event signature with other cluster events

venimir-ssv committed:
OperatorMigratedToETH event added

venimir-ssv committed:
Remove timestamps from DAO root events

venimir-ssv committed:
fix: align ClusterBalanceUpdated indexing with other cluster events

venimir-ssv committed:
Merge pull request #325 from ssvlabs/fix/cluster-balance-updated-indexing

fix: align ClusterBalanceUpdated indexing with other cluster events

yurii-ssv committed:
feat: add effective balance to getter

yurii-ssv committed:
feat: add cluster liquidation upon update

yurii-ssv committed:
Merge pull request #326 from ssvlabs/feat/cluster-balance-get-and-liquidate

Update cluster balance getter to include effective balance and add liquidation check during effective balance update

venimir-ssv committed:
eb added to migrate event, operator default eth fee fixed

venimir-ssv committed:
comment cleanup

venimir-ssv committed:
refactor: centralize validator register/remove flows in SSVClusters (#327)

* refactor: centralize validator register/remove flows in SSVClusters

venimir-ssv committed:
Merge pull request #329 from ssvlabs/fix/add-eb-to-event

eb added to migrate event, operator default eth fee fixed

yurii-ssv committed:
Migrate to hardhat v3 (#328)

* feat: update to hardhat v3, add scripts, track deployments

venimir-ssv committed:
operator version removed, migrate operator refactored

venimir-ssv committed:
operator constants refactored

venimir-ssv committed:
Allow migrating liquidated SSV clusters without double-counting operators

venimir-ssv committed:
Ref/operator version (#330)

* operator version removed, migrate operator refactored

venimir-ssv committed:
Merge branch 'fix/liquidated-ssv-cluster-migration' into feat/eth-eb-merge

yurii-ssv committed:
feat: add eb minimum balance check (#333)

yurii-ssv committed:
Update balance getters and handle eb amount in gwei (#331)

* feat: update balance getters with eb and eth logic
* feat: handle eb balance in ETH units in function and event params

yurii-ssv committed:
fix: unit32 eb / operator struct (#334)

venimir-ssv committed:
operator version removed

venimir-ssv committed:
Merge pull request #337 from ssvlabs/fix/remove-operator-version

operator version removed

yurii-ssv committed:
Reduce vunits scaling (#336)

* feat: reduce vunits scaling factor
* feat: remove vunits from `ClusterBalanceUpdated`

yurii-ssv committed:
Export abis and store them in the repo (#335)

* feat: ignore libs, interfaces and dependencies in abi exporter
* feat: remove abis folder from gitignore
* feat: add main contracts abis to the repo

yurii-ssv committed:
Change eb return data types (#338)

* feat: change return type to uint32
* feat: change precision data type to uint32

mtabasco committed:
feat:staking contract + cSSV WIP

mtabasco committed:
Merge branch 'feat/eth-eb-merge' into feat/staking-contract

yurii-ssv committed:
Feat/hoodi dev deployment (#339)

* chore: deployed addresses + abis

mtabasco committed:
feat:update deploy script, remove withdrawNetworkEarnings, optimizations

yurii-ssv committed:
fix: module names

mtabasco committed:
feat:updateable cooldownPeriod, remove address(0) check on cSSV

venimir-ssv committed:
docs: update ETH migration changelog from base to current HEAD (#341)

- Updated changelog to reflect all changes from commit a2e968f to bff3aed
- Added comprehensive documentation for SSV Staking Contract feature
- Added detailed file changes for staking-related contracts (SSVStaking, CSSVToken)
- Updated commit history with Phase 6: Staking Contract commits
- Updated statistics: 65 files changed, +15,708/-13,239 lines
- Removed obsolete references to withdrawNetworkEarnings (ETH version)
- Added staking migration path, security considerations, and testing recommendations
- Documented ETH earnings distribution through staking contract

yurii-ssv committed:
feat: add support for custom initializers

yurii-ssv committed:
chore: deployed addresses and abis

mtabasco committed:
sync base branch

mtabasco committed:
Staking contract (#340)

* feat:staking contract + cSSV token

mtabasco committed:
feat:WIP - data structures, base delagation logic

yurii-ssv committed:
chore: force compile on abi saving

yurii-ssv committed:
chore: change eb data type to uint32

yurii-ssv committed:
feat: adjust `ClusterMigratedToETH ` to uint32

yurii-ssv committed:
chore: return param names for `getBalance`

yurii-ssv committed:
chore: update abis

yurii-ssv committed:
chore: rename eb in interface

yurii-ssv committed:
chore: abis

yurii-ssv committed:
chore: update deployments

venimir-ssv committed:
Set default quorumBps on deployment and update docs

venimir-ssv committed:
Initialize quorum and default oracles via config during deployment

venimir-ssv committed:
Merge pull request #343 from ssvlabs/chore/default-quorum-bps

Set default quorumBps on deployment and update docs

venimir-ssv committed:
Handle delegation weight on cSSV transfers and move quorum/oracle config into initializer

yurii-ssv committed:
Feature/merkle leaf double hash (#344)

* feat: do double keccak on proofs

* chore: redeploy clusters module

venimir-ssv committed:
Merge pull request #345 from ssvlabs/chore/staking-delegation-transfer

Handle delegation weight on cSSV transfers and move quorum/oracle con…

venimir-ssv committed:
Optimize delegation updates

venimir-ssv committed:
Merge pull request #346 from ssvlabs/chore/optimize-cssv-delegation-transfer

Optimize delegation updates

yurii-ssv committed:
feat: update eth burn rate calculation (#347)

{[Network Fee] + [Σ Operator Fees]} X {[Total Effective Balance (ETH)] / 32]

venimir-ssv committed:
merge

yurii-ssv committed:
fix: remove duplicated event

yurii-ssv committed:
Refactor balance getters (#348)

* feat: remove eb from `getBalance` getters

* feat: introduce `getEffectiveBalance` getter

* chore: update abis

olegshmuelov commented on contracts/modules/SSVDAO.sol:
```diff
@@ -75,4 +86,104 @@ contract SSVDAO is ISSVDAO {
         emit OperatorMaximumFeeUpdated(maxFee);
     }
+
+    function commitRoot(bytes32 merkleRoot, uint64 blockNum) external override {
+        StorageEB storage seb = SSVStorageEB.load();
+        StorageStaking storage s = SSVStorageStaking.load();
+
+        uint32 oracleId = s.oracleIdOf[msg.sender];
+        if (oracleId == 0) revert NotOracle();
+
+        // Enforce monotonicity - new block must be greater than last
+        if (blockNum <= seb.latestCommittedBlock) {
+            revert StaleBlockNumber();
+        }
+
+        // Ensure block is not in the future
+        if (blockNum > block.number) {
+            revert FutureBlockNumber();
+        }
+
+        // block and root combined to keep block-root proposal tied together
+        bytes32 commitmentKey = keccak256(abi.encodePacked(blockNum, merkleRoot));
+        
+        if (seb.hasVoted[commitmentKey][oracleId]) revert AlreadyVoted();
+        seb.hasVoted[commitmentKey][oracleId] = true;
+
+        uint256 weight = s.oracleWeights[oracleId];
+        seb.rootCommitments[commitmentKey] += weight;
+
+        uint256 accumulatedWeight = seb.rootCommitments[commitmentKey];
+        uint256 totalSupply = ICSSVToken(s.cssv).totalSupply();
+
+        uint256 threshold = (totalSupply * s.quorumBps) / 10000;
+
```
The `commitRoot` function has no access control - anyone can call it and vote on root commitments. This is a critical security vulnerability that allows:

1. **Unauthorized voting**: Any address can vote, not just authorized oracles
2. **Vote manipulation**: Same address can vote multiple times (no duplicate check)
3. **Root injection**: Attacker votes 3 times -> malicious root is committed

Suggestion: Add oracle authorization and duplicate vote prevention:
```solidity
// In StorageEB struct, add:
mapping(address => bool) authorizedOracles;
mapping(bytes32 => mapping(address => bool)) hasVoted;

function commitRoot(bytes32 merkleRoot, uint64 blockNum) external override {
    StorageEB storage seb = SSVStorageEB.load();

    // Check oracle authorization
    if (!seb.authorizedOracles[msg.sender]) {
        revert NotAuthorizedOracle();
    }

    // ... existing validation ...

    bytes32 commitmentKey = keccak256(abi.encodePacked(blockNum, merkleRoot));

    // Prevent duplicate votes
    if (seb.hasVoted[commitmentKey][msg.sender]) {
        revert AlreadyVoted();
    }
    seb.hasVoted[commitmentKey][msg.sender] = true;

    seb.rootCommitments[commitmentKey] += 1;
    // ... rest of logic
}
```

olegshmuelov commented on contracts/libraries/ProtocolLib.sol:
```diff
@@ -12,13 +13,25 @@ library ProtocolLib {
+        sp.ethNetworkFee = fee.shrink();
+    }
+
+    function updateNetworkFeeSSV(StorageProtocol storage sp, uint256 fee) internal {
+        updateDAOEarnings(sp);
+
+        sp.networkFeeIndex = currentNetworkFeeIndexSSV(sp);
         sp.networkFeeIndexBlockNumber = uint32(block.number);
         sp.networkFee = fee.shrink();
     }
```
The `updateNetworkFeeSSV` function calls `updateDAOEarnings(sp)` which updates **ETH** fields (`ethDaoBalance`, `ethDaoIndexBlockNumber`) instead of **SSV** fields (`daoBalance`, `daoIndexBlockNumber`).

**Impact:** When SSV network fee is changed:
1. `daoBalance` and `daoIndexBlockNumber` are NOT settled
2. The fee change is applied
3. Future calls to `networkTotalEarningsSSV()` will calculate earnings from `daoIndexBlockNumber` to now using the NEW fee
4. This incorrectly applies the new fee to the period BEFORE the fee change

Compare with `updateDAOSSV()` at line 79 which correctly calls `updateDAOEarningsSSV()`.

Suggestion: Fix the function call:

```solidity
function updateNetworkFeeSSV(StorageProtocol storage sp, uint256 fee) internal {
    updateDAOEarningsSSV(sp);  // Changed from updateDAOEarnings(sp)

    sp.networkFeeIndex = currentNetworkFeeIndexSSV(sp);
    sp.networkFeeIndexBlockNumber = uint32(block.number);
    sp.networkFee = fee.shrink();
}
```

venimir-ssv replied:
@olegshmuelov good catch we have addressed it here https://github.com/ssvlabs/ssv-network/pull/350

olegshmuelov commented on contracts/modules/SSVStaking.sol:
```diff
@@ -0,0 +1,389 @@
+        uint256 claimable = s.accrued[msg.sender];
+        if (claimable == 0) revert NothingToClaim();
+
+        // Round down to precision supported by protocol storage
```
Users with accrued rewards below `DEDUCTED_DIGITS` (10,000,000 wei) cannot claim their rewards at all:

```solidity
uint256 payout = claimable - (claimable % DEDUCTED_DIGITS);
if (payout == 0) revert NothingToClaim();  // Reverts if claimable < DEDUCTED_DIGITS
```

If a user has 9,999,999 wei in rewards, `payout = 0` and the function reverts. This dust is stuck forever if the user unstakes all cSSV.

Suggestion: Allow users to claim any non-zero amount, even if below precision threshold:

```solidity
if (claimable == 0) revert NothingToClaim();
uint256 payout = claimable;  // Claim full amount
s.accrued[msg.sender] = 0;
// Transfer payout
```

olegshmuelov commented on contracts/modules/SSVClusters.sol:
```diff
@@ -357,4 +289,502 @@ contract SSVClusters is ISSVClusters {
+    }
+
+    function _emitClusterBalanceUpdated(
+        address clusterOwner,
+        uint64[] calldata operatorIds,
+        uint64 blockNum,
+        uint32 eb,
+        uint64 newVUnits,
+        Cluster memory cluster
+    ) internal {
+        emit ClusterBalanceUpdated(clusterOwner, operatorIds, blockNum, eb, cluster);
+    }
```
The `_emitClusterBalanceUpdated` function has an unused parameter `newVUnits`:

olegshmuelov commented on contracts/modules/SSVDAO.sol:
```diff
@@ -75,4 +86,104 @@ contract SSVDAO is ISSVDAO {
+        emit QuorumUpdated(quorum);
+    }
+
+    function setOracleTimingConfig(
```
The `setOracleTimingConfig` function appears unnecessary since oracle timing configuration is managed by the oracle itself

olegshmuelov commented on contracts/libraries/OperatorLib.sol:
```diff
@@ -5,34 +5,131 @@ import "../interfaces/ISSVNetworkCore.sol";
+        uint32 currentBlock = uint32(block.number);
+        uint64 blockDiffEthFee = (currentBlock - operator.ethSnapshot.block) * operator.ethFee;
+
+        uint64 vUnits = seb.operatorEthVUnits[operatorId];
+        if (vUnits == 0 && operator.ethValidatorCount > 0) {
+            vUnits = operator.ethValidatorCount * VUNITS_PRECISION;
+        }
```
I think there is a bug in vUnits fallback logic causing non-deterministic operator earnings.

The fallback in `updateSnapshot` is all-or-nothing at the operator level:

```solidity
uint64 vUnits = seb.operatorEthVUnits[operatorId];
if (vUnits == 0 && operator.ethValidatorCount > 0) {
    vUnits = operator.ethValidatorCount * VUNITS_PRECISION;  // fallback
}
```

**Problem:** Once ANY cluster calls `updateClusterBalance`, the fallback is disabled for ALL clusters.

**Example:**
- Operator has 2 clusters (A, B), each with 1 validator
- Before updates: `operatorEthVUnits=0`, fallback gives **20,000 vUnits**
- Cluster A updates: `operatorEthVUnits=10,000`, fallback disabled
- After: operator earns on **10,000 vUnits** (50% loss!)
- Cluster B's 10,000 vUnits contribution is lost until B also updates

**Impact:**
1. Operator earnings drop significantly during transition period
2. Earnings depend on ORDER of cluster updates (non-deterministic)
3. Operators are systematically underpaid until all their clusters update EB

**Suggested fix:** Initialize `operatorEthVUnits` with the fallback value when first cluster updates:

```solidity
function _updateOperatorVUnits(...) internal {
    uint64 storedVUnits = seb.clusterEB[clusterId].vUnits;

    // If this is first explicit vUnits for this operator, initialize with fallback
    if (seb.operatorEthVUnits[operatorId] == 0) {
        Operator storage op = SSVStorage.load().operators[operatorId];
        seb.operatorEthVUnits[operatorId] = op.ethValidatorCount * VUNITS_PRECISION;
    }

    // Now apply delta as before...
}
```

olegshmuelov commented on contracts/modules/SSVClusters.sol:
```diff
@@ -357,4 +289,502 @@ contract SSVClusters is ISSVClusters {
+                }
+            }
+
+            uint256 n = operatorIds.length;
+            for (uint256 i; i < n; ++i) {
+                uint64 opId = operatorIds[i];
+                if (version == CoreLib.VERSION_ETH) seb.operatorEthVUnits[opId] -= vUnitsCluster;
+                else seb.operatorVUnits[opId] -= vUnitsCluster;
+            }
```
In `_executeLiquidation`, the code subtracts `vUnitsCluster` from operator vUnits without checking if the operator has sufficient vUnits:

```solidity
if (version == CoreLib.VERSION_ETH) seb.operatorEthVUnits[opId] -= vUnitsCluster;
else seb.operatorVUnits[opId] -= vUnitsCluster;
```

If an operator's vUnits were reduced elsewhere (e.g., through another cluster's liquidation happening concurrently), this could underflow. While Solidity 0.8.x has built-in overflow/underflow protection, the transaction would revert unexpectedly, potentially blocking liquidations.

olegshmuelov commented on contracts/modules/SSVClusters.sol:
```diff
@@ -357,4 +289,502 @@ contract SSVClusters is ISSVClusters {
+        emit ClusterMigratedToETH(msg.sender, operatorIds, msg.value, ssvBalance, effectiveBalance, cluster);
+    }
+
+    function updateClusterBalance(
+        uint64 blockNum,
+        address clusterOwner,
+        uint64[] calldata operatorIds,
+        Cluster memory cluster,
+        uint32 effectiveBalance,
+        bytes32[] calldata merkleProof
+    ) external override {
+        UpdateCtx memory ctx;
+        StorageData storage s = SSVStorage.load();
+
+        (ctx.clusterId, ctx.version) = cluster.validateHashedCluster(clusterOwner, operatorIds, s);
+        ctx.clusterOwner = clusterOwner;
+        ctx.blockNum = blockNum;
+        ctx.effectiveBalance = effectiveBalance;
+        ctx.merkleProof = merkleProof;
+
+        _updateClusterBalanceInternal(operatorIds, cluster, ctx);
+    }
```
should we update balance for liquidated clusters?

venimir-ssv committed:
updateDAOEarningsSSV fixed (#350)

olegshmuelov committed:
feat: add getCommittedRoot view function for oracle commit status (#352)

yurii-ssv committed:
Feature/hotfixes and tests (#351)

* fixes based on feedback and code review
* tests

mtabasco committed:
Fix/effective balance precision loss (#360)

* fix: prevent effective balance precision loss in roundtrip conversion

yurii-ssv committed:
chore: abis

yurii-ssv committed:
chore: add lib to abis excude

yurii-ssv committed:
chore: increase hardhat gas limit for local tests

venimir-ssv committed:
staking unit tests init

venimir-ssv committed:
storage data tests added

venimir-ssv committed:
Readme added

yurii-ssv committed:
feat: allow multiple unstake requests

venimir-ssv committed:
Merge pull request #361 from ssvlabs/feat/ssv-staking-unit

staking unit tests init

yurii-ssv committed:
feat: simplify requests tracking

Added an array to track requests instead of heavt weight sets and maps

yurii-ssv committed:
fix: remove redundant import

venimir-ssv committed:
gas profile and gas stats commands added

venimir-ssv committed:
post initial stake test track gas added

venimir-ssv committed:
test: add gas tracking for existing ETH cluster registration

venimir-ssv committed:
add gas tracking for missing groups

yurii-ssv committed:
feat: simplify balance check

yurii-ssv committed:
feat: remove total requested calculation

yurii-ssv committed:
feat: optimize the calculation loop

yurii-ssv committed:
feat: add max requests validation

yurii-ssv committed:
chore: remove unused errors

yurii-ssv committed:
feat: sync tests with latest changes

yurii-ssv committed:
Merge branch 'feat/eth-eb-merge' into feature/partial-unstake

yurii-ssv committed:
Merge pull request #362 from ssvlabs/feature/partial-unstake

Allow multiple unstake requests

venimir-ssv committed:
Merge pull request #363 from ssvlabs/feat/gas-profiling

gas profile and gas stats commands added

mtabasco committed:
feat: split SSVClusters, adapt tests, deployment scripts (#364)

yurii-ssv committed:
chore: update abis

venimir-ssv committed:
gas track added

venimir-ssv committed:
Add SSV liquidate gas groups and align staking withdrawal tests (#365)

* Add SSV liquidate gas groups and align staking withdrawal tests

* gas track added to ci

* actions fixed

venimir-ssv committed:
fix(harness): remove duplicate withdrawal getters

venimir-ssv committed:
Fix/removed operator (#366)

* fix: update cluster operators refactor (#367)
